<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Defines a new exception type."><title>create_exception in pyo3 - Rust</title><script> if (window.location.protocol !== "file:") document.write(`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2">`)</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-e935ef01ae1c1829.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="pyo3" data-themes="" data-resource-suffix="" data-rustdoc-version="1.78.0 (9b00956e5 2024-04-29) (Arch Linux rust 1:1.78.0-1)" data-channel="1.78.0" data-search-js="search-42d8da7a6b9792c2.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-4c98445ec4002617.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-12cf3b4f4f9dc36d.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../pyo3/index.html">pyo3</a><span class="version">0.21.2</span></h2></div><div class="sidebar-elems"></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../pyo3/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Macro <a href="index.html">pyo3</a>::<wbr><a class="macro" href="#">create_exception</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/pyo3/exceptions.rs.html#238-263">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><span class="macro">macro_rules!</span> create_exception {
    (<span class="macro-nonterminal">$module</span>: expr, <span class="macro-nonterminal">$name</span>: ident, <span class="macro-nonterminal">$base</span>: ty) =&gt; { ... };
    (<span class="macro-nonterminal">$module</span>: expr, <span class="macro-nonterminal">$name</span>: ident, <span class="macro-nonterminal">$base</span>: ty, <span class="macro-nonterminal">$doc</span>: expr) =&gt; { ... };
}</pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Defines a new exception type.</p>
<h2 id="syntax"><a class="doc-anchor" href="#syntax">§</a>Syntax</h2>
<ul>
<li><code>module</code> is the name of the containing module.</li>
<li><code>name</code> is the name of the new exception type.</li>
<li><code>base</code> is the base class of <code>MyError</code>, usually <a href="exceptions/struct.PyException.html" title="struct pyo3::exceptions::PyException"><code>PyException</code></a>.</li>
<li><code>doc</code> (optional) is the docstring visible to users (with <code>.__doc__</code> and <code>help()</code>) and
accompanies your error type in your crate’s documentation.</li>
</ul>
<h2 id="examples"><a class="doc-anchor" href="#examples">§</a>Examples</h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>pyo3::prelude::<span class="kw-2">*</span>;
<span class="kw">use </span>pyo3::create_exception;
<span class="kw">use </span>pyo3::exceptions::PyException;

<span class="macro">create_exception!</span>(my_module, MyError, PyException, <span class="string">"Some description."</span>);

<span class="attr">#[pyfunction]
</span><span class="kw">fn </span>raise_myerror() -&gt; PyResult&lt;()&gt; {
    <span class="kw">let </span>err = MyError::new_err(<span class="string">"Some error happened."</span>);
    <span class="prelude-val">Err</span>(err)
}

<span class="attr">#[pymodule]
</span><span class="kw">fn </span>my_module(m: <span class="kw-2">&amp;</span>Bound&lt;<span class="lifetime">'_</span>, PyModule&gt;) -&gt; PyResult&lt;()&gt; {
    m.add(<span class="string">"MyError"</span>, m.py().get_type_bound::&lt;MyError&gt;())<span class="question-mark">?</span>;
    m.add_function(<span class="macro">wrap_pyfunction!</span>(raise_myerror, m)<span class="question-mark">?</span>)<span class="question-mark">?</span>;
    <span class="prelude-val">Ok</span>(())
}</code></pre></div>
<p>Python code can handle this exception like any other exception:</p>
<div class="example-wrap"><pre class="language-python"><code>from my_module import MyError, raise_myerror

try:
    raise_myerror()
except MyError as e:
    assert e.__doc__ == &#39;Some description.&#39;
    assert str(e) == &#39;Some error happened.&#39;
</code></pre></div></div></details></section></div></main></body></html>